finalitygadget/finalitygadget.go

```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
 	}, nil
 }
 
-func (fg *FinalityGadget) handleBlock(ctx context.Context, latestFinalizedHeight uint64) error {
-	for height := fg.lastProcessedHeight + 1; height <= latestFinalizedHeight; height++ {
+type blockResult struct {
+	height      uint64
+	block       *types.Block
+	isFinalized bool
+	err         error
+}
+
+func (fg *FinalityGadget) processBlocksTillHeight(ctx context.Context, latestHeight uint64) error {
+	for startHeight := fg.lastProcessedHeight + 1; startHeight <= latestHeight; {
 		select {
 		case <-ctx.Done():
 			return nil
 		default:
-			if height > math.MaxInt64 {
-				return fmt.Errorf("block height %d exceeds maximum int64 value", height)
+			// Calculate batch start and end heights
+			endHeight := startHeight + fg.batchSize - 1
+			if endHeight > latestHeight {
+				endHeight = latestHeight
 			}
-			block, err := fg.queryBlockByHeight(int64(height))
-			if err != nil {
-				return fmt.Errorf("error getting block at height %d: %w", height, err)
+			fg.logger.Info("Processing batch of blocks", zap.Uint64("start_height", startHeight), zap.Uint64("end_height", endHeight))
+
+			// Create batch of blocks to check in parallel
+			results := make(chan blockResult, endHeight-startHeight+1)
+			var wg sync.WaitGroup
+
+			// Query batch in parallel
+			for height := startHeight; height <= endHeight; height++ {
+				if height > math.MaxInt64 {
+					return fmt.Errorf("block height %d exceeds maximum int64 value", height)
```

Comment:
why do we need this check? if there a type conversion below?

but anyway shouldn't we check `endHeight` instead of `startHeight`? if yes, shouldn't we just check the passed in argument `latestHeight` at the first line of the function instead of inside the loop?

Reply:
this is old code - i think its because eth rpc uses i64 for block heights

Reply:
yes you can see there is a type conversion later. but my point is this check is incorrect/incomplete.

we can remove it here and add this to the top:
```
func (fg *FinalityGadget) processBlocksTillHeight(ctx context.Context, latestHeight uint64) error {
	if latestHeight > math.MaxInt64 {
		return fmt.Errorf("block height %d exceeds maximum int64 value", height)
	}
```